Doc 12549 post ga tidy (#911)
couchbaselabs/docs-couchbase-lite: modules/ROOT/pages/_partials/commons/common-vector-search.adoc

== Key Concepts of Vector Search in Couchbase Lite 

When xref:{param-module}:working-with-vector-search.adoc[working with Vector Search], you should be aware of the core concepts below. 

=== About Vector Embeddings

Vector embeddings represent the output of a Machine Learning (ML) model as an array of numbers to capture semantic or contextual relationships between data points.
This representation encodes how a ML model understands the input or inputs provided to it, based on how the model was initially trained and the internal structure of the model.
When a model considers the features of a given input as similar, the distance between the vector embeddings will be short.
Vector embeddings are stored within embedded vector indexes.

The current supported formats for Vector embeddings are:

* An array of 32-bit floats.

* A Base64 string that encodes a https://en.wikipedia.org/wiki/Endianness[Little-endian] array of 32-bit floats.

=== About Vector Indexes

Vector Indexes are used to store and manage vector representations of content in the form of vector embeddings.
You can use Vector Indexes to efficiently retrieve vectors, similar to a target vector.

> Before use, a Vector Index needs to be trained to compute the centroids and parameters for encoding the vectors.

Reviewer: So, you mention the "minimum-training-size' configuration in the paragraph below. Is this something that the customer sets? Where is it and how do they set it?

If it is indeed under the control of the user, I'd suggest changing to something like:

Before use, a Vector Index needs to be trained to compute the centroids and parameters for encoding the vectors. You configure the number of vectors required in the index for training to take place using the minimum training size option[[[specify where and how]]. 

If the database contains enough vectors, Index training begins automatically when the first Vector Search query is performed on the database. (If the database does not contain the required number of vectors, an error message appears in the log.)

Author: Done! I've opted to point to the working with doc for context-relevant snippets displaying how you can configure training sizes and to point to the configuration table so users can view possible configuration options.

NOTE: If there are insufficient vectors in the database for training, a warning message will be logged, indicating the required number of vectors.

Training is configured to commence automatically on the first Vector Search query when there are enough vectors to be trained based on the `minimum-training-size` configuration.
Be aware that vector index training can affect query performance.
If a query is executed against the index before it is trained, a full scan of the vectors will be performed. 

=== About Lazy Vector Indexes

IMPORTANT: Lazy index is not an automatic process, you will need to manually schedule the index updates. 

Lazy vector indexes (lazy index) is Couchbase Lite specific functionality that updates indexes asynchronously, satisfying the following use cases:

* Documents have been added to the application by the end-user with no available Machine Learning (ML) model to generate the vectors.

> You can use lazy indexing to schedule updates to the index of said documents when a ML model is available. 

Reviewer: I would change "said documents" to "such docments"

Author: Done!

* The remote ML model used stops working or has intermittent availability, causing a failed update.

> With lazy index, you can skip documents that fail to update and schedule the index process to be retried at a later date.

Reviewer: With lazy indexing, you can skip documents that fail to update and schedule the index process at a later date.

Author: Done!

* When to update the index.

* The number of vectors to update to the index.

* Whether to cancel or skip certain indexes when the model is unavailable or has failed.

NOTE: Updating in lazy index is an independent process from saving document operations.

See here for examples of how to xref:{param-module}:working-with-vector-search.adoc#create-a-lazy-vector-index[use lazy index in your applications].

Lazy indexing is an alternate approach to using the standard predictive model with regular vector indexes which handle the indexing process automatically. 
The table below compares the two processes.

.Regular Vector Indexes vs Lazy Vector Indexes
[cols ="3*"]
|===
|Feature |Regular Index |Lazy Index

|Update when documents are changed
|image:ROOT:yes.png[]
|image:ROOT:no.png[]

|Update when documents are deleted or purged
|image:ROOT:yes.png[]
|image:ROOT:yes.png[]

|The application has control when to update the index
|image:ROOT:no.png[]
|image:ROOT:yes.png[]

|The application can skip updating the index
|image:ROOT:no.png[]
|image:ROOT:yes.png[]

|Is an asynchronous process
|image:ROOT:no.png[]
|image:ROOT:yes.png[]

|===

=== About Vector Encoding

> Vector encoding reduces the size of Vectors index by algorithmic compression.

Reviewer: I think vectors should be lower case, unless when used as a proper noun (Vector Search )

Author: I agree, changed.

> You can configure the Vector Encoding in Couchbase Lite depending on your application's needs.

Reviewer: Change "depending on your" to "to address your"

Author: Done

The number of bits per component can be set to 4, 6, or 8 bits.
The default setting in Couchbase Lite is 8 bits Scalar Quantizer or SQ-8.

* Product Quantizer - This reduces the number of dimensions and bits per dimension. 
It splits the vectors into multiple subspaces and performing scalar quantization on each space independently before compression.
This can produce higher quality results than Scalar Quantization at the cost of greater complexity.

NOTE: Quantizers are algorithmic processes that map input values from a larger set to output values in a smaller set, common quantization processes can include operations such as rounding and truncation.

=== About Centroids

Centroids are vectors that function as the center point of a vector cluster within the data set.
Each vector is then associated to the vector it is closest to by https://en.wikipedia.org/wiki/K-means_clustering[k-means clustering.]
Each Centroid is contained within a bucket along with its associated vectors, the greater amount of Centroids, the greater the potential accuracy of the model.
This will also incur a longer indexing time.

> Choosing Centroids in vector search involves trade-offs that can impact clustering effectiveness and search efficiency. The initial selection of Centroids, the number chosen, and their sensitivity to high dimensionality and outliers affect the quality of vector clustering.

Reviewer: i-cap Vector Search when referring to the feature.

Author: Good spot! Corrected throughout

=== About Vector Embeddings

Vector embeddings represent the output of a Machine Learning (ML) model as an array of numbers to capture semantic or contextual relationships between data points.
This representation encodes how a ML model understands the input or inputs provided to it, based on how the model was initially trained and the internal structure of the model.
When a model considers the features of a given input as similar, the distance between the vector embeddings will be short.
Vector embeddings are stored within embedded vector indexes.

The current supported formats for Vector embeddings are:

* An array of 32-bit floats.

* A Base64 string that encodes a https://en.wikipedia.org/wiki/Endianness[Little-endian] array of 32-bit floats.

=== About Vector Indexes

Vector Indexes are used to store and manage vector representations of content in the form of vector embeddings.
You can use Vector Indexes to efficiently retrieve vectors, similar to a target vector.
Before use, a Vector Index needs to be trained to compute the centroids and parameters for encoding the vectors.

> You can configure both the minimum and maximum training sizes for your vector index by setting the relevant parameters in your vector index configuration, for an example of how you can do so, see xref:{param-module}:working-with-vector-search.adoc#create-a-vector-index[create a vector index].

Reviewer: I'd split into two sentence..."in your vector index configuration. For an example of how you can do so...."

Author: Done

> Training is configured to commence automatically on the first Vector Search query when there are enough vectors to be trained based on the `minimum-training-size` configuration.

Reviewer: Sorry--one more comment and question.

First, I'd say "begin" rather than "commence," as it is more in line with our casual voice and tone.

My real question, though, is...as a user, do I configure training to begin automatically? Or is this pre-configured by us for the user?

Author: Condensed and clarified, thanks for the input!

The table below compares the two processes.

.Regular Vector Indexes vs Lazy Vector Indexes
[cols ="3*"]
|===
|Feature |Regular Index |Lazy Index

|Update when documents are changed
|image:ROOT:yes.png[]
|image:ROOT:no.png[]

|Update when documents are deleted or purged
|image:ROOT:yes.png[]
|image:ROOT:yes.png[]

|The application has control when to update the index
|image:ROOT:no.png[]
|image:ROOT:yes.png[]

|The application can skip updating the index
|image:ROOT:no.png[]
|image:ROOT:yes.png[]

|Is an asynchronous process
|image:ROOT:no.png[]
|image:ROOT:yes.png[]

|===

=== About Vector Encoding

Vector encoding reduces the size of the vectors index by algorithmic compression.
You can configure the Vector Encoding in Couchbase Lite to address your application's needs.

This vector encoding compression reduces disk space required and I/O time during indexing and queries, but greater compression can result in inaccurate results in distance calculations.

Vector Search for Couchbase Lite supports the following encoding algorithms:

> * None - This will return the highest quality results but at high performance and disk space costs

Reviewer 2: Add full stop for consistency with other bullet points in this list

```suggestion
* None - This will return the highest quality results but at high performance and disk space costs.
```

* Scalar Quantizer - This reduces the number of bits used for each number in a vector. 
The number of bits per component can be set to 4, 6, or 8 bits.
The default setting in Couchbase Lite is 8 bits Scalar Quantizer or SQ-8.

* Product Quantizer - This reduces the number of dimensions and bits per dimension. 
It splits the vectors into multiple subspaces and performing scalar quantization on each space independently before compression.
This can produce higher quality results than Scalar Quantization at the cost of greater complexity.

NOTE: Quantizers are algorithmic processes that map input values from a larger set to output values in a smaller set, common quantization processes can include operations such as rounding and truncation.

=== About Centroids

Centroids are vectors that function as the center point of a vector cluster within the data set.
Each vector is then associated to the vector it is closest to by https://en.wikipedia.org/wiki/K-means_clustering[k-means clustering.]

> Each Centroid is contained within a bucket along with its associated vectors, the greater amount of Centroids, the greater the potential accuracy of the model.

Reviewer 2: New sentence. Also not sure about "amount"

```suggestion
Each Centroid is contained within a bucket along with its associated vectors. The greater the number of Centroids, the greater the potential accuracy of the model.
```

> This will also incur a longer indexing time.

Reviewer 2: Just checking, does this mean that more Centroids will incur a longer indexing time?

```suggestion
However, a greater number of Centroids will incur a longer indexing time.
```

Author: It does, I'll reword for clarity.

Choosing Centroids in Vector Search involves trade-offs that can impact clustering effectiveness and search efficiency. The initial selection of Centroids, the number chosen, and their sensitivity to high dimensionality and outliers affect the quality of vector clustering.

> The general guideline for the amount of Centroids is approximately the square root of the number of documents.

Reviewer 2: ```suggestion
The general guideline for the optimum number of Centroids is approximately the square root of the number of documents.
```

=== About Probes

> The number of Probes refers to the maximum number of Centroid buckets that the search algorithm will to check to look for similar vectors to a given query vector.

Reviewer 2: typo, delete

```suggestion
The number of Probes refers to the maximum number of Centroid buckets that the search algorithm will check to look for similar vectors to a given query vector.
```

You can change the number of Probes by altering the value of the `NumProbes` variable xref:{param-module}:working-with-vector-search.adoc#create-a-vector-index[shown in the following example].
Couchbase recommends that when setting a custom number of probes, the number should be at least 8 or 0.5% the number of Centroids used.

=== About Dimensions

> Vector dimensions describes the amount of numbers in a given vector embedding, commonly known as its' width.

Reviewer 2: typo

```suggestion
Vector dimensions describes the amount of numbers in a given vector embedding, commonly known as its width.
```

> The greater the amount of dimensions, the greater accuracy of results, this also results in greater compute and memory costs and a increase in the latency of the search.

Reviewer 2: Reword slightly, like the comment above about Centroids, to make it clear that we're talking the advantages and tradeoffs of increasing the number of dimensions. Something like

```suggestion
The greater the number of dimensions, the greater the accuracy of results.
However, a greater number of dimensions also results in greater compute and memory costs and a increase in the latency of the search.
```

Vector embeddings are stored within embedded vector indexes.

The current supported formats for Vector embeddings are:

* An array of 32-bit floats.

* A Base64 string that encodes a https://en.wikipedia.org/wiki/Endianness[Little-endian] array of 32-bit floats.

=== About Vector Indexes

Vector Indexes are used to store and manage vector representations of content in the form of vector embeddings.
You can use Vector Indexes to efficiently retrieve vectors, similar to a target vector.
Before use, a Vector Index needs to be trained to compute the centroids and parameters for encoding the vectors.

You can configure both the minimum and maximum training sizes for your vector index by setting the relevant parameters in your vector index configuration.
For an example of how you can do so, see xref:{param-module}:working-with-vector-search.adoc#create-a-vector-index[create a vector index].

Couchbase Lite Vector Search initiates training on the first Vector Search query automatically when the amount of vectors to be trained satisfies the `minimum-training-size` configuration.

> If the database does not contain the required number of vectors, an error message will be logged indicating the required number of vectors.

Reviewer: Did you decide not to bullet this out? That's fine. I would change "amount of vectors" to "number of vectors"

Author: I decided to keep it as a full sentence. Even though it's still a little wordy it boils down to one conditions. Made the change to number of!
